### ISSUE TITLE
XJC class visibility too restrictive

### ISSUE THREAD

**Issue by bertramn** [eclipse-ee4j/jaxb-ri#1314]
Is it possible to relax the visibility of core XJC code model classes? codemodel is generally useful but with package private constructors on things like [JAnnotationUse](https://github.com/eclipse-ee4j/jaxb-ri/blob/master/jaxb-ri/codemodel/codemodel/src/main/java/com/sun/codemodel/JAnnotationUse.java#L39) this library is rather useless.

Here is another example where we cannot effectively use the classes to write custom XJC compiler extensions. If we want to replace a field on a class model we cannot set its parent as [CPropertyInfo.setParent](https://github.com/eclipse-ee4j/jaxb-ri/blob/master/jaxb-ri/xjc/src/main/java/com/sun/tools/xjc/model/CPropertyInfo.java#L119) is only visible at package level.

**Comment by phax** [comment]
Hi, https://github.com/phax/jcodemodel is a possibility to use codemodel outside of the context of JAXB - maybe that also solves your issues.

**Comment by bertramn** [comment]
Not really solving my problem, I am trying to write portable XJC compiler plugins. If I were to use your project, I'd have to write tones of adapter classes and half way through get stuck because some of the sun classes are also final.

**Comment by phax** [comment]
Yeah okay - for XJC plugins, I don't have a solution. I am using it only outside of XJC...

**Comment by laurentschoelens** [comment]
@bertramn could you explain is the use-case of `new JAnnotationUse()` outside of annotating a class / param / ... ?

> Here is another example where we cannot effectively use the classes to write custom XJC compiler extensions. If we want to replace a field on a class model we cannot set its parent as [CPropertyInfo.setParent](https://github.com/eclipse-ee4j/jaxb-ri/blob/master/jaxb-ri/xjc/src/main/java/com/sun/tools/xjc/model/CPropertyInfo.java#L119) is only visible at package level.

I was in same situation as you but you should go with `cclassInfo.addProperty(...)` which will do the `setParent` part. I'll submit a `addProperty(index, ...)` to `cclassinfo`
